sim/network.go

```diff
@@ -61,24 +61,76 @@ func (n *Network) AddParticipant(p gpbft.Receiver) {
 	n.participants[p.ID()] = p
 }
 
+////// Network interface
+
+func (n *Network) NetworkFor(signer gpbft.Signer, id gpbft.ActorID) *networkFor {
+	return &networkFor{
+		ParticipantID: id,
+		Signer:        signer,
+		Network:       n,
+	}
+}
+
+type networkFor struct {
+	ParticipantID gpbft.ActorID
+	Signer        gpbft.Signer
+	*Network
+}
+
+// Log implements tagging Tracer interface
+func (nf *networkFor) Log(format string, args ...any) {
+	nf.Network.log(TraceLogic, "P%d "+format, append([]any{nf.ParticipantID}, args...)...)
+}
+
+func (nf *networkFor) RequestBroadcast(msg *gpbft.GMessage) {
+	nf.log(TraceSent, "P%d ↗ %v", msg.Sender, msg)
+	for _, dest := range nf.participantIDs {
+		latencySample := time.Duration(0)
+		if dest != msg.Sender {
+			latencySample = nf.latency.Sample(nf.Time(), msg.Sender, dest)
+		}
+
+		nf.queue.Insert(
+			&messageInFlight{
+				source:    msg.Sender,
+				dest:      dest,
+				payload:   *msg,
+				deliverAt: nf.clock.Add(latencySample),
+			})
+	}
+}
+
 func (n *Network) NetworkName() gpbft.NetworkName {
 	return n.networkName
 }
 
-func (n *Network) Broadcast(msg *gpbft.GMessage) {
-	n.log(TraceSent, "P%d ↗ %v", msg.Sender, msg)
+func (n *Network) Broadcast(sender gpbft.ActorID, msg *gpbft.GMessage) {
+	n.broadcast(sender, msg, false)
+}
+
+func (n *Network) BroadcastSynchronous(sender gpbft.ActorID, msg *gpbft.GMessage) {
+	n.broadcast(sender, msg, true)
+}
+
+func (n *Network) broadcast(sender gpbft.ActorID, msg *gpbft.GMessage, synchronous bool) {
+	n.log(TraceSent, "P%d ↗ %v", sender, msg)
 	for _, dest := range n.participantIDs {
-		if dest != msg.Sender {
-			now := n.Time()
-			delay := n.latency.Sample(now, msg.Sender, dest)
-			n.queue.Insert(
-				&messageInFlight{
-					source:    msg.Sender,
-					dest:      dest,
-					payload:   *msg,
-					deliverAt: now.Add(delay),
-				})
+		if dest == sender {
+			continue
+		}
```

Comment:
I think we want to deliver locally here too, with the same logic as `RequestBroadcast`. I don't think we need the sender parameter at all. I think the only difference here is the synchronous mode. You should be able to delegate to this from `RequestBroadcast` too.